core/thread/id.go

```diff
@@ -184,84 +171,138 @@ func uvError(read int) error {
 	}
 }
 
+func (i ID) Valid() bool {
+	data := i.Bytes()
+	err := validateIDData(data)
+	return err == nil
+}
+
+func getVersion(data []byte) (uint64, int, error) {
+	vers, n := binary.Uvarint(data)
+	if err := uvError(n); err != nil {
+		return 0, 0, err
+	}
+	return vers, n, nil
+}
+
+func validateIDData(data []byte) error {
+	vers, n, err := getVersion(data)
+	if err != nil {
+		return err
+	}
+
+	if vers != V1 {
+		return fmt.Errorf("expected 1 as the id version number, got: %d", vers)
+	}
+
+	variant, cn := binary.Uvarint(data[n:])
+	if err := uvError(cn); err != nil {
+		return err
+	}
+
+	if variant != uint64(Raw) && variant != uint64(AccessControlled) {
+		return fmt.Errorf("expected Raw or AccessControlled as the id variant, got: %d", variant)
+	}
+
+	id := data[n+cn:]
+	if len(id) == 0 {
+		return fmt.Errorf("expected random id bytes but there are none")
+	}
+
+	return nil
+}
+
 // UnmarshalBinary is equivalent to Cast(). It implements the
 // encoding.BinaryUnmarshaler interface.
 func (i *ID) UnmarshalBinary(data []byte) error {
-	casted, err := Cast(data)
+	id, err := Cast(data)
 	if err != nil {
 		return err
 	}
-	i.str = casted.str
+	*i = id
 	return nil
 }
 
 // UnmarshalText is equivalent to Decode(). It implements the
 // encoding.TextUnmarshaler interface.
 func (i *ID) UnmarshalText(text []byte) error {
-	decodedID, err := Decode(string(text))
+	id, err := Decode(string(text))
 	if err != nil {
 		return err
 	}
-	i.str = decodedID.str
+	*i = id
 	return nil
 }
 
 // Version returns the ID version.
 func (i ID) Version() uint64 {
-	return V1
+	version, _, err := getVersion(i.Bytes())
+	if err != nil {
+		panic("error getting version: " + err.Error())
+	}
+	return version
 }
 
 // Variant returns the variant of an ID.
 func (i ID) Variant() Variant {
-	_, n := uvarint(i.str)
-	variant, _ := uvarint(i.str[n:])
+	_, n := uvarint(string(i))
+	variant, _ := uvarint(string(i)[n:])
```

Comment:
This seems like it could be similar to `Version()` and panic if there were an error reading the variant byte, but it uses this `uvarint()` method I don't really understand. Just tell me if we should leave this or make it more like `Version()`.

This also doesn't verify it's a known `Variant`, but calling `Valid()` does check that.

Reply:
Yeah let's just leave this for now. Maybe we can revisit when adding the ACL stuff to ID.